File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/Greedy/JoinRopes.swift
import Foundation

class Heap {
    private var size : Int // Number of elements in Heap
    private var arr : [Int] // The Heap array
    private var isMinHeap : Bool

    init(_ isMin : Bool) {
        self.arr = []
        self.size = 0
        self.isMinHeap = isMin
    }

    init(_ array : inout [Int], _ isMin : Bool) {
        self.size = array.count
        self.arr = array
        self.isMinHeap = isMin

		// Build Heap operation over array
		var i : Int = (self.size / 2)
		while (i >= 0) {
			self.percolateDown(i)
			i -= 1
		}
    }

    func compare(_ first : Int, _ second : Int) -> Bool {
        if (self.isMinHeap) {
            return (self.arr[first] > self.arr[second])
        } else {
            return (self.arr[first] < self.arr[second])
        }
    }

    // Other Methods.
    func percolateDown(_ parent : Int) {
        let lChild : Int = 2 * parent + 1
        let rChild : Int = lChild + 1
        var child : Int = -1
        if (lChild < self.size) {
            child = lChild
        }
        if (rChild < self.size && self.compare(lChild,rChild)) {
            child = rChild
        }
        if (child != -1 && self.compare(parent,child)) {
            let temp : Int = self.arr[parent]
            self.arr[parent] = self.arr[child]
            self.arr[child] = temp
            self.percolateDown(child)
        }
    }

    func percolateUp(_ child : Int) {
        let parent : Int = (child - 1) / 2
        if (parent >= 0 && self.compare(parent,child)) {
            let temp : Int = self.arr[child]
            self.arr[child] = self.arr[parent]
            self.arr[parent] = temp
            self.percolateUp(parent)
        }
    }

    func isEmpty() -> Bool
    {
        return (self.size == 0)
    }

    func length() -> Int
    {
        return self.size
    }

    func peek() -> Int
    {
        if (self.isEmpty())
        {
            print("Heap empty exception.")
			return 0
        }
        return self.arr[0]
    }

    func add(_ value : Int) {
		self.size += 1
		self.arr.append(value)
        self.percolateUp(self.size - 1)
    }

    func remove() -> Int {
        if (self.isEmpty()) {
            print("Heap empty exception.")
			return 0
        }
        let value : Int = self.arr[0]
        self.arr[0] = self.arr[self.size - 1]
        self.size -= 1
		self.arr.removeLast()
        self.percolateDown(0)
        return value
    }

    func printHeap() {
        print("Heap : ",terminator: "")
		var i : Int = 0
		while (i < self.size) {
			print(String(self.arr[i]), terminator: " ")
			i += 1
		}
        print()
    }

    func delete(_ value : Int) -> Bool
    {
		var i : Int = 0
		while (i < self.size) {
			if (self.arr[i] == value) {
				self.arr[i] = self.arr[self.size - 1]
				self.size -= 1
				self.arr.removeLast()
				self.percolateUp(i)
				self.percolateDown(i)
				return true
			}
			i += 1
		}
        return false
    }
}

func joinRopes(_ ropes : inout [Int], _ size : Int) -> Int {
    ropes.sort()
    var i : Int = 0, j : Int = size - 1
    while (i < j) {
        let temp : Int = ropes[i]
        ropes[i] = ropes[j]
        ropes[j] = temp
        i += 1
        j -= 1
    }
    
    var total : Int = 0
    var value : Int = 0
    var index : Int
    var length : Int = size
    while (length >= 2) {
        value = ropes[length - 1] + ropes[length - 2]
        total += value
        index = length - 2
        while (index > 0 && ropes[index - 1] < value) {
            ropes[index] = ropes[index - 1]
            index -= 1
        }
        ropes[index] = value
        length -= 1
    }
    return total
}

func joinRopes2(_ ropes : inout [Int], _ size : Int) -> Int {
    let pq = Heap(true)
    var i : Int = 0
    while (i < size) {
        pq.add(ropes[i])
        i += 1
    }
    
    var total : Int = 0
    var value : Int = 0
    while (pq.length() > 1) {
        value = pq.remove()
        value += pq.remove()
        pq.add(value)
        total += value
    }
    return total
}

// Testing code.
var ropes : [Int] = [4, 3, 2, 6]
print("Total : " + String(joinRopes( &ropes,ropes.count)))
var rope2 : [Int] = [4, 3, 2, 6]
print("Total : " + String(joinRopes2( &rope2,rope2.count)))
/* 
Total : 29
Total : 29
 */

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/Greedy/JoinRopes.ts
class PriorityQueue<T> {
    static CAPACITY: number = 32;
    size: number;
    arr: Array<T>;
    compare: (a: T, b: T) => boolean;

    public constructor(cmp: (a: T, b: T) => boolean) {
        this.compare = cmp;
        this.arr = new Array<T>(PriorityQueue.CAPACITY);
        this.size = 0;
    }

    percolateDown(parent: number) {
        let lChild: number = 2 * parent + 1;
        let rChild: number = lChild + 1;
        let child: number = -1;
        let temp: T;
        if (lChild < this.size) {
            child = lChild;
        }
        if (rChild < this.size && this.compare(this.arr[lChild], this.arr[rChild])) {
            child = rChild;
        }
        if (child !== -1 && this.compare(this.arr[parent], this.arr[child])) {
            temp = this.arr[parent];
            this.arr[parent] = this.arr[child];
            this.arr[child] = temp;
            this.percolateDown(child);
        }
    }

    percolateUp(child: number) {
        let parent: number = Math.floor((child - 1) / 2);
        let temp: T;
        if (parent < 0) {
            return;
        }
        if (this.compare(this.arr[parent], this.arr[child])) {
            temp = this.arr[child];
            this.arr[child] = this.arr[parent];
            this.arr[parent] = temp;
            this.percolateUp(parent);
        }
    }

    public add(value: T) {
        if (this.size === this.arr.length) {
            this.doubleSize();
        }

        this.arr[this.size++] = value;
        this.percolateUp(this.size - 1);
    }

    private doubleSize() {
        let old: Array<T> = this.arr;
        let newSize = this.size * 2;
        this.arr = new Array<T>(newSize);
        /* arraycopy */
        var size = this.size;
        for (let i = 0; i < size; i++) {
            this.arr[i] = old[i];
        }
    }

    public remove(): T {
        if (this.isEmpty()) {
            throw new Error("IllegalStateException");
        }
        let value: T = this.arr[0];
        this.arr[0] = this.arr[this.size - 1];
        this.size--;
        this.percolateDown(0);
        return value;
    }

    public PrintTree() {
        console.log(this.arr);
    }

    public isEmpty(): boolean {
        return this.size === 0;
    }

    public length(): number {
        return this.size;
    }

    public peek(): T {
        if (this.isEmpty()) {
            throw new Error("IllegalStateException");
        }
        return this.arr[0];
    }
}


function joinRopes(ropes: number[], size: number): number {
    ropes.sort((a, b) => b - a);

    let total = 0;
    let value = 0;
    let length = size;

    while (length >= 2) {
        value = ropes[length - 1] + ropes[length - 2];
        total += value;
        let index = length - 2;
        while (index > 0 && ropes[index - 1] < value) {
            ropes[index] = ropes[index - 1];
            index--;
        }
        ropes[index] = value;
        length--;
    }

    console.log("Total:", total);
    return total;
}

function joinRopes2(ropes: number[], size: number): number {
    const pq: PriorityQueue<number> = new PriorityQueue<number>((a, b) => a > b);

    for (let i = 0; i < size; i++) {
        pq.add(ropes[i]);
    }

    let total = 0;
    while (pq.length() > 1) {
        const value = pq.remove() + pq.remove();
        pq.add(value);
        total += value;
    }

    console.log("Total:", total);
    return total;
}

/* Testing Code */
const ropes: number[] = [4, 3, 2, 6];
joinRopes(ropes, ropes.length);

const ropes2: number[] = [4, 3, 2, 6];
joinRopes2(ropes2, ropes2.length);

/*
Total: 29
Total: 29
*/
